### PR TITLE
Improve test covereage of driver discovery.

### PR DESCRIPTION
**Pull request by danielballan**
This PR adds a test driver that implements entrypoints and adds some
test coverage with and without the package scan activated. It also fixes two
bugs in the new features introduced by #236.

### CODE REVIEW

**Comment by martindurant** [comment]
A few fixes needed here

**Comment by danielballan** [comment]
Ha. Turns out it helps to commit the new files. :-D

**Comment by danielballan** [comment]
...But there is still an issue, some coupling between tests via the configuration file.

**Comment by danielballan** [comment]
@martindurant Something very odd is happening here. One of the tests is causing the user config file at `~/.intake/conf.yaml` to be changed. Specifically, it adds a `drivers` block with some entries that confuse subsequent tests and cause them to fail.

Before:
```
auth:                                                                                                                                                                                         
  cls: intake.auth.base.BaseAuth                                                                                                                                                              
cache_dir: /home/dallan/.intake/cache                                                                                                                                                         
cache_disabled: false                                                                                                                                                                         
cache_download_progress: true                                                                                                                                                                 
catalog_path: []                                                                                                                                                                                                                                                                                              
logging: INFO                                                                                                                                                                                 
persist_path: /home/dallan/.intake/persisted                                                                                                                                                  
port: 5000      
```

After:
```
auth:                                                                                                                                                                                         
  cls: intake.auth.base.BaseAuth                                                                                                                                                              
cache_dir: /home/dallan/.intake/cache                                                                                                                                                         
cache_disabled: false                                                                                                                                                                         
cache_download_progress: true                                                                                                                                                                 
catalog_path: []                                                                                                                                                                              
drivers:                                                                                                                                                                                      
  foo: intake_foo.FooPlugin                                                                                                                                                                   
  some_test_driver: driver_with_entrypoints.SomeTestDriver                                                                                                                                    
logging: INFO                                                                                                                                                                                 
persist_path: /home/dallan/.intake/persisted                                                                                                                                                  
port: 5000      
```

In 2def020 I added a (temporary) hook to read my config file after every test runs and try to isolate which test is the culprit. Weirdly, I find that the file has been changed after `test_text_export`.

```
intake/source/tests/test_text.py::test_complex_bytes[pars2-bz2] PASSED                                                                                                                 [ 84%]
intake/source/tests/test_text.py::test_complex_bytes[pars3-None] PASSED                                                                                                                [ 84%]
intake/source/tests/test_text.py::test_complex_bytes[pars3-gzip] PASSED                                                                                                                [ 85%]
intake/source/tests/test_text.py::test_complex_bytes[pars3-bz2] PASSED                                                                                                                 [ 85%]
intake/source/tests/test_text.py::test_text_persist PASSED                                                                                                                             [ 85%]
intake/source/tests/test_text.py::test_text_export FAILED                                                                                                                              [ 86%]
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

item = <Function test_text_export>

    @pytest.hookimpl(hookwrapper=True)
    def pytest_runtest_call(item):
        "Verify that no tests modify the user's config file."
        yield
        with open('/home/dallan/.intake/conf.yaml') as file:
>           assert 'drivers' not in yaml_load(file)
E           AssertionError: assert 'drivers' not in {'auth': {'cls': 'intake.auth.base.BaseAuth'}, 'cache_dir': '/home/dallan/.intake/cache', 'cache_disabled': False, 'cache_download_progress': True, ...}
E            +  where {'auth': {'cls': 'intake.auth.base.BaseAuth'}, 'cache_dir': '/home/dallan/.intake/cache', 'cache_disabled': False, 'cache_download_progress': True, ...} = yaml_load(<_io.TextIOWrapper name='/home/dallan/.intake/conf.yaml' mode='r' encoding='UTF-8'>)

intake/conftest.py:193: AssertionError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[5] > /home/dallan/Repos/bnl/intake/intake/conftest.py(193)pytest_runtest_call()
-> assert 'drivers' not in yaml_load(file)
```

But a check inside the _body_ of `test_text_export` itself does _not_ trip. Have you encountered this weird behavior before?

**Comment by danielballan** [comment]
The "action at a distance" here is still a mystery, but [this line](https://github.com/intake/intake/pull/386/files#diff-f90316ac37ee73d2e7c6afac071e8759R35) fixed it.

**Comment by martindurant** [comment]
Certainly seems like a reasonable line to add. I suppose we are reusing something from one test to another.

**Comment by danielballan** [comment]
I think this is good to go. Worth getting these fixes in.

File: intake/source/tests/plugin_searchpath/driver_with_entrypoints/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+class SomeTestDriver:
+    ...
```
Comment (martindurant) [inline intake/source/tests/plugin_searchpath/driver_with_entrypoints/__init__.py]: Does this get loaded? Should be `pass`?

Comment (danielballan) [reply intake/source/tests/plugin_searchpath/driver_with_entrypoints/__init__.py]: It does get loaded. Ellipsis is a valid substitute for `pass` in py3.

Comment (martindurant) [reply intake/source/tests/plugin_searchpath/driver_with_entrypoints/__init__.py]: Still learning....

**Comment by martindurant** [comment]
Glad to have you working on this, thanks.

File: intake/source/tests/plugin_searchpath/driver_with_entrypoints/__init__.py (hunk 1 above)
Comment (danielballan) [reply intake/source/tests/plugin_searchpath/driver_with_entrypoints/__init__.py]: Me too! :-D

**Pull request merged**